pkg/cni/types/types.go

```diff
@@ -0,0 +1,108 @@
+/*
+ Copyright 2023, NVIDIA CORPORATION & AFFILIATES
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+     http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/containernetworking/cni/pkg/types"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/cni/k8sclient"
+)
+
+const (
+	DefaultConfDir    = "/etc/cni/net.d/nv-ipam.d"
+	DefaultKubeConfig = DefaultConfDir + "/" + "nv-ipam.kubeconfig"
+	DefaultDataDir    = "/var/lib/cni/nv-ipam"
+	DefaultLogFile    = "/var/log/nv-ipam-cni.log"
+
+	HostLocalDataDir = "state/host-local"
+	K8sNodeNameFile  = "k8s-node-name"
+)
+
+type IPAMConf struct {
+	types.IPAM
+
+	// PoolName is the name of the pool to be used to allocate IP
+	PoolName   string `json:"poolName"`
+	Kubeconfig string `json:"kubeconfig,omitempty"`
+	DataDir    string `json:"dataDir,omitempty"`
+	ConfDir    string `json:"confDir,omitempty"`
+	LogFile    string `json:"logFile,omitempty"`
+	LogLevel   string `json:"logLevel,omitempty"`
+
+	// internal configuration
+	NodeName  string
+	K8sClient *kubernetes.Clientset
```

Comment:
I think it better if we create sidecare daemon to call GetNode() periodically and save the information into file what do you think?

Reply:
I think it adds un-necessary complexity for a problem we don't know exists and that we dont know it needs solving.

today multus performs calls to k8s on every pod start and whereabouts (regardless of leader election) does so as well so i dont think its a problem for this phase.

Reply:
also an alternative can be to store the annot in file with time stamp,
then only fetch from k8s if deadline exceeded.

but if we end up needing to deal with it, probably the sidecar is nicer